src/NServiceBus.Core/Recoverability/SecondLevelRetries/SecondLevelRetriesBehavior.cs

```diff
@@ -15,16 +15,40 @@ namespace NServiceBus
 
     class SecondLevelRetriesBehavior : Behavior<TransportReceiveContext>
     {
-        public SecondLevelRetriesBehavior(IPipelineBase<RoutingContext> dispatchPipeline, SecondLevelRetryPolicy retryPolicy, BusNotifications notifications, string localAddress)
+        public SecondLevelRetriesBehavior(IPipelineBase<RoutingContext> dispatchPipeline, SecondLevelRetryPolicy retryPolicy, BusNotifications notifications, string localAddress, SlrStatusStorage slrStatusStorage)
         {
             this.dispatchPipeline = dispatchPipeline;
             this.retryPolicy = retryPolicy;
             this.notifications = notifications;
             this.localAddress = localAddress;
+            this.slrStatusStorage = slrStatusStorage;
         }
 
         public override async Task Invoke(TransportReceiveContext context, Func<Task> next)
         {
+            Exception exception;
+            var uniqueMessageId = PipelineInfo.Name + context.Message.MessageId;
+
+            if (slrStatusStorage.TryGetException(uniqueMessageId, out exception))
+            {
+                slrStatusStorage.ClearException(uniqueMessageId);
```

Comment:
I that's the expected behavior. So if exception happens when SLR performs it's logic that exception is caught by faults behavior which handles the message afterwards. In other words I think that we should clear the exception in the beginning because whatever happens next (either successful dispatch to timeout queue, exception on dispatch or giving up) the message should not return to SLR.